src/EntryPoint.sol

```diff
@@ -649,6 +672,57 @@ contract EntryPoint is
         }
     }
 
+    /// @dev Loops over the `encodedPreOps` and does the following for each sub UserOp:
+    /// - Check that the eoa is indeed the eoa of the parent UserOp.
+    /// - If there are any sub UserOp in a sub UserOp, recurse.
+    /// - Validate the sub UserOp.
+    /// - Check and increment the nonce of the sub UserOp.
+    /// - Call the Delegation with `executionData` in the sub UserOp, using the ERC7821 batch-execution mode.
+    ///   If the call fails, revert.
+    /// - Emit an {UserOpExecuted} event.
+    function _handlePreOps(address eoa, uint256 simulationFlags, bytes[] calldata encodedPreOps)
+        internal
+        virtual
+    {
+        for (uint256 i; i < encodedPreOps.length; ++i) {
+            UserOp calldata u = _extractUserOp(encodedPreOps[i]);
+            if (eoa != u.eoa) revert InvalidPreOpEOA();
+
+            // The order is exactly the same as `selfCallPayVerifyCall537021665`:
+            // Recurse -> Verify -> Increment nonce -> Call eoa.
+            if (u.encodedPreOps.length != 0) _handlePreOps(eoa, simulationFlags, u.encodedPreOps);
+
+            (bool isValid, bytes32 keyHash) = _verify(u);
+            if (!isValid) if (simulationFlags & 1 == 0) revert PreOpVerificationError();
+
+            LibNonce.checkAndIncrement(_getEntryPointStorage().nonceSeqs[eoa], u.nonce);
+
+            // This part is same as `selfCallPayVerifyCall537021665`. We simply inline to save gas.
+            bytes memory data = LibERC7579.reencodeBatchAsExecuteCalldata(
+                hex"01000000000078210001", // ERC7821 batch execution mode.
+                u.executionData,
+                abi.encode(keyHash) // `opData`.
+            );
+            // This part is slightly different from `selfCallPayVerifyCall537021665`.
+            // It always reverts on failure.
+            assembly ("memory-safe") {
+                mstore(0x00, 0) // Zeroize the return slot.
+                if iszero(call(gas(), eoa, 0, add(0x20, data), mload(data), 0x00, 0x20)) {
+                    // If this is a simulation via `simulateFailed`, bubble up the whole revert.
```

Comment:
What should the gas limit be here? We cannot forward everything here. My suggestion was that we cap each PreOp to 200K gas (or whatever the most expensive key mgmt operation is?) so the relay can predict how much they can get grieved for in the worst case? The relay can also just have a simple PreOps whitelist to ensure they don't call any unsafe method.

Reply:
We can cap it to 200k gas.